File: nlm_floating_point/PDCCH/ps_nr_pdcch_coreset_extraction.m
function coreset_sym_rbs = ps_nr_pdcch_coreset_extraction(rx_data,nSym,symb_offset,rb_offset,rbs,kssb)
  
 cp_first_symbol = 88;
 cp_other_symbol = 72;
 fs              = 30.72e+6;
 
 cp_length_type_0 = [88,72,72];
 cp_length_type_1 = [72,72,72];
 
 fftSize            = 1024;
 n_re_rb            = 12;
 ssb_rbs            = 20;
 kssb_scs           = 15e+3;
 
 start_rb_second_half  = ssb_rbs/2+rb_offset; % This is the offset from the DC
 start_re_second_half  = fftSize - start_rb_second_half*n_re_rb+1;
 end_rb_second_half    = (rbs - start_rb_second_half)*n_re_rb;
 
 coreset_sym_rbs = [];
 
 crb_alignment_freq_shift = kssb_scs*kssb;
 
 rx_data1            = rx_data.*exp(1j.*(2.*pi./fs).*crb_alignment_freq_shift.*(0:length(rx_data)-1).');
  
 if(symb_offset == 0)
   sym_offset = 1;
   for n = 1:nSym
     cp_length       = cp_length_type_0(n);
     jitt_gaurd      = cp_length/4;
     phase_derot_jitt_gaurd   = exp(j*2*pi*(cp_length-jitt_gaurd).*[0:fftSize-1].'./fftSize);
     rx_symb_fft     = fft(rx_data1(sym_offset + jitt_gaurd + (0:fftSize-1))).*phase_derot_jitt_gaurd;
     re_first_half   = rx_symb_fft(start_re_second_half:end);
     re_second_half  = rx_symb_fft(1:end_rb_second_half);
     coreset_sym_rbs = [coreset_sym_rbs;[re_first_half;re_second_half]];
     sym_offset      = sym_offset+cp_length+fftSize;      
   end
  else
   sym_offset = (cp_length_type_0(1)+fftSize) + (symb_offset-1)*(cp_length_type_0(2)+fftSize)+1;
   for n = 1:nSym
     cp_length       = cp_length_type_1(n);
     jitt_gaurd      = cp_length/4;
     phase_derot_jitt_gaurd   = exp(j*2*pi*(cp_length-jitt_gaurd).*[0:fftSize-1].'./fftSize);
     rx_symb_fft     = fft(rx_data1(sym_offset + jitt_gaurd + (0:fftSize-1))).*phase_derot_jitt_gaurd;
     re_first_half   = rx_symb_fft(start_re_second_half:end);
     re_second_half  = rx_symb_fft(1:end_rb_second_half);
     coreset_sym_rbs = [coreset_sym_rbs;[re_first_half;re_second_half]];
     sym_offset     = sym_offset+cp_length+fftSize;     
   end
 end
 
end

File: nlm_application/examples/nmm/lib/ps_nr_pdcch_coreset_extraction.c
#include <sys/types.h>
#include "armral.h"
#include <string.h>
#include "ps_nr_pdcch.h"
#include <rte_memcpy.h>

void ps_nr_pdcch_coreset_extraction(armral_cmplx_int16_t *inSym,uint8_t nSym,uint8_t symb_offset,uint8_t rb_offset,uint8_t rbs,uint8_t kssb,armral_cmplx_int16_t *outSym, void * scratchBuf)
{
 
 armral_fft_direction_t dir = ARMRAL_FFT_FORWARDS; // We are taking FFT here. 
 armral_fft_plan_t *p = NULL;
 armral_fft_create_plan_cs16(&p, 1024, dir);
  
 uint8_t cp_length;
 uint8_t cp_length_type_0[3] = {88,72,72};
 uint8_t cp_length_type_1[3] = {72,72,72};
 
 uint16_t fftSize            = 1024;
 uint8_t n_re_rb             = 12;
 uint8_t ssb_rbs             = 20;
    
 uint16_t start_rb_second_half  = (ssb_rbs>>1) + rb_offset; // This is the offset from the DC
 uint16_t start_re_second_half  = fftSize - start_rb_second_half*n_re_rb;
 uint16_t end_re_first_half     = (rbs - start_rb_second_half)*n_re_rb;

 armral_cmplx_int16_t *rx_symb_fft = (armral_cmplx_int16_t *)scratchBuf;
        
 if(symb_offset == 0)
 {	
 	
 	uint32_t sym_offset = 0;

  	for(uint8_t n = 0;n<nSym; n++)
  	{	
    	cp_length               = cp_length_type_0[n];
    	uint32_t sym_start_loc  = sym_offset+cp_length;
        
        // Insert FFT here // 
        armral_fft_execute_cs16(p,inSym + sym_start_loc,rx_symb_fft);
    	
    	for(uint16_t re_idx = start_re_second_half; re_idx<fftSize; re_idx++)
		{   
			rx_symb_fft[re_idx].re >>= 5; // FFT will lead to 5 bit of bit growth. Which may saturate the data at peak power
            rx_symb_fft[re_idx].im >>= 5;
    		rte_memcpy(outSym++,&rx_symb_fft[re_idx],sizeof(armral_cmplx_int16_t));
		}
        for(uint16_t re_idx = 0; re_idx<end_re_first_half; re_idx++)
		{   
			rx_symb_fft[re_idx].re >>= 5; // FFT will lead to 5 bit of bit growth. Which may saturate the data at peak power
            rx_symb_fft[re_idx].im >>= 5;
    		rte_memcpy(outSym++,&rx_symb_fft[re_idx],sizeof(armral_cmplx_int16_t));
		} 		
    	sym_offset += cp_length+fftSize;    	    
    }
  } 
  else
  {	
   
   uint16_t sym_offset = cp_length_type_0[0]+fftSize + (cp_length_type_0[1]+fftSize)*(symb_offset-1);
   
   for(uint8_t n=0;n<nSym; n++)
   {	
    	cp_length       = cp_length_type_1[n];
    	uint32_t sym_start_loc  = sym_offset+cp_length;
     
        // Insert FFT here //
     	armral_fft_execute_cs16(p,inSym + sym_start_loc,rx_symb_fft);
     
    	for(uint16_t re_idx = start_re_second_half; re_idx<fftSize; re_idx++)
		{   
			rx_symb_fft[re_idx].re >>= 5; // FFT will lead to 5 bit of bit growth. Which may saturate the data at peak power
            rx_symb_fft[re_idx].im >>= 5;
    		rte_memcpy(outSym++,&rx_symb_fft[re_idx],sizeof(armral_cmplx_int16_t));
		}
    	for(uint16_t re_idx = 0; re_idx<end_re_first_half; re_idx++)
		{
			rx_symb_fft[re_idx].re >>= 5; // FFT will lead to 5 bit of bit growth. Which may saturate the data at peak power
            rx_symb_fft[re_idx].im >>= 5;
    		rte_memcpy(outSym++,&rx_symb_fft[re_idx],sizeof(armral_cmplx_int16_t));
		}  
    	symb_offset += cp_length+fftSize;     
   }
  }

  armral_fft_destroy_plan_cs16(&p);


}  